test/RouterUniV2.js

```diff
@@ -127,12 +146,28 @@ describe('Router [UniV2]', async function () {
 
         it('uniswap', async function () {
```

Comment:
I found the cheaper way to use UnversalRouter:
<img width="520" alt="Screenshot" src="https://github.com/user-attachments/assets/61baffc2-2c5c-4400-924e-b4b59ccb9a72">

```
const { ethers, getChainId } = require('hardhat');
const { ether, constants, trim0x } = require('@1inch/solidity-utils');
const { RoutePlanner, CommandType, UniswapTrade, SwapRouter } = require('@uniswap/universal-router-sdk');
const { Trade: V2Trade, Route: RouteV2, Pair, computePairAddress } = require('@uniswap/v2-sdk');
const { TradeType, CurrencyAmount, Token, Ether, Percent } = require('@uniswap/sdk-core');
const { Trade: RouterTrade } = require('@uniswap/router-sdk');

...

it.only('uniswap', async function () {
    const {
        tokens,
        uniswapUniversalRouter,
        settings: { gasUsedTableRow, amount },
        addr1,
    } = await loadFixture(initContractsWithCaseSettings);

    async function getPair(tokenA, tokenB) {
        const V2_FACTORY = '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f'
        const V2_ABI = ["function getReserves() view returns (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast)"];

        const pairAddress = computePairAddress({ factoryAddress: V2_FACTORY, tokenA, tokenB })
        const contract = await ethers.getContractAt(V2_ABI, pairAddress)
        const { reserve0, reserve1 } = await contract.getReserves()
        const [token0, token1] = tokenA.sortsBefore(tokenB) ? [tokenA, tokenB] : [tokenB, tokenA] // does safety checks
        return new Pair(CurrencyAmount.fromRawAmount(token0, reserve0.toString()), CurrencyAmount.fromRawAmount(token1, reserve1.toString()))
    }
    function buildTrade(trades) {
        return new RouterTrade({
          v2Routes: trades
            .filter((trade) => trade instanceof V2Trade)
            .map((trade) => ({
              routev2: trade.route,
              inputAmount: trade.inputAmount,
              outputAmount: trade.outputAmount,
            })),
            v3Routes: [],
            mixedRoutes: [],
        //   v3Routes: trades
        //     .filter((trade) => trade instanceof V3Trade)
        //     .map((trade) => ({
        //       routev3: trade.route as RouteV3<Currency, Currency>,
        //       inputAmount: trade.inputAmount,
        //       outputAmount: trade.outputAmount,
        //     })),
        //   mixedRoutes: trades
        //     .filter((trade) => trade instanceof MixedRouteTrade)
        //     .map((trade) => ({
        //       mixedRoute: trade.route as MixedRouteSDK<Currency, Currency>,
        //       inputAmount: trade.inputAmount,
        //       outputAmount: trade.outputAmount,
        //     })),
          tradeType: trades[0].tradeType,
        })
    }

    const ETHER = Ether.onChain(await getChainId())
    const trade = new V2Trade(
        new RouteV2(
            [
                await getPair(new Token(1, tokens.WETH.target, 18, 'WETH', 'Wrapped Ether'), new Token(1, tokens.USDC.target, 6, 'USDC', 'USDC')),
                await getPair(new Token(1, tokens.USDC.target, 6, 'USDC', 'USDC'), new Token(1, tokens.DAI.target, 18, 'DAI', 'DAI')),
            ],
            ETHER,
            new Token(1, tokens.DAI.target, 18, 'DAI', 'DAI'),
        ),
        CurrencyAmount.fromRawAmount(ETHER, amount.toString()),
        TradeType.EXACT_INPUT,
    );
    const methodParameters = SwapRouter.swapCallParameters(new UniswapTrade(buildTrade([trade]), {slippageTolerance: new Percent(5, 100)}));
    const tx = await addr1.sendTransaction({
        to: await uniswapUniversalRouter.getAddress(),
        value: methodParameters.value,
        data: methodParameters.calldata,
    })

    gasUsedTable.addElementToRow(gasUsedTableRow, ProtocolKey.UNISWAP, (await tx.wait()).gasUsed);
});
```

Maybe it is cheaper in another cases. This is very rough and dirty code; there are certainly many opportunities for unification and refactoring. This code is simply an example of achieving a cheaper transaction than is currently possible in test.

Reply:
The [example](https://github.com/Uniswap/sdks/blob/92b765bdf2759e5e6639a01728a96df81efbaa2b/sdks/universal-router-sdk/test/uniswapTrades.test.ts#L76-L90) where I got it

Reply:
Maybe we can realise it with `createCommand` [method](https://github.com/Uniswap/sdks/blob/2eee6b49ecb803aacc4cfee84e62f10804d37809/sdks/universal-router-sdk/src/utils/routerCommands.ts#L149)

Reply:
It seems I may have misled you a bit. While router in your code does indeed use the `Permit2` contract, it doesn’t utilize the signing and permission functionality. Instead, it simply uses the `Permit2` contract as a proxy via the `approve` method. In other words, your approach is acceptable because it interacts with the `Permit2` contract, even though it doesn’t leverage the `Permit` technology itself. It also seems that the router creators intended for it to work this way.

However, we should investigate why I was able to perform a swap with `srcToken=ETH` at a lower cost. It’s possible that there's no need to explicitly wrap ETH, as the router might automatically handle transactions with a non-zero value by default.

Reply:
It seems that the code I suggested doesn't actually impact the gas usage, as the current value is also 155k (you need just fix it in readme). This might be due to some changes in the pool's state, which previously caused it to be more expensive. Anyway, you can mark this conversation as resolved. 🙃

<img width="459" alt="Снимок экрана 2024-08-10 в 20 24 50" src="https://github.com/user-attachments/assets/aa808c4e-c047-4479-bbd7-ad70b7308ad9">